updatehub-sdk/src/listener.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2018, 2019, 2020 O.S. Systems Sofware LTDA
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Error, Result};
+use async_ctrlc::CtrlC;
+use async_std::{
+    io::BufReader,
+    os::unix::net::{UnixListener, UnixStream},
+    prelude::*,
+};
+use slog::*;
+use std::{collections::HashMap, net::Shutdown, process};
+
+const SDK_TRIGGER_FILENAME: &str =
+    "/usr/share/updatehub/state-change-callbacks.d/10-updatehub-sdk-statechange-trigger";
+const SOCKET_PATH: &str = "/run/updatehub-statechange.sock";
+
+type StateChangeCallback = Box<dyn Fn(State)>;
+type ErrorCallback = Box<dyn Fn(String)>;
+
+#[derive(Default)]
+pub struct StateChangeListener {
+    listeners: HashMap<StateID, Vec<StateChangeCallback>>,
+    error_handlers: Vec<ErrorCallback>,
+}
+
+#[derive(PartialEq, Eq, Hash)]
+pub enum StateID {
+    Download,
+    Install,
+    Reboot,
+    Error,
+}
+
+pub struct State {
+    pub id: StateID,
+    pub conn: UnixStream,
+}
+
+impl State {
+    pub fn cancel(&mut self) {
+        self.conn.write_all(b"cancel");
+    }
+
+    pub fn proceed(&self) {}
+}
+
+impl StateChangeListener {
+    pub fn new() -> Self {
+        StateChangeListener { listeners: HashMap::new(), error_handlers: Vec::new() }
+    }
+
+    pub fn on(&mut self, state: StateID, f: StateChangeCallback) {
+        match self.listeners.get_mut(&state) {
+            Some(callbacks) => {
+                callbacks.push(Box::new(f));
+            }
+            None => {
+                self.listeners.insert(state, vec![Box::new(f)]);
+            }
+        }
+    }
+
+    pub fn on_error(&mut self, f: ErrorCallback) {
+        self.error_handlers.push(f);
+    }
+
+    pub async fn listen(&self) -> Result<()> {
+        let file = std::path::Path::new(&SDK_TRIGGER_FILENAME);
+        if !file.exists() {
+            let decorator = slog_term::TermDecorator::new().build();
+            let drain = slog_term::FullFormat::new(decorator).build().fuse();
+            let drain = slog_async::Async::new(drain).build().fuse();
+
+            let log = slog::Logger::root(drain, o!());
+            info!(
+                log,
+                "WARNING: updatehub-sdk-statechange-trigger not found on {}", SDK_TRIGGER_FILENAME
+            );
+        }
+        std::fs::remove_file(SOCKET_PATH)?;
+        let listener = UnixListener::bind(SOCKET_PATH).await?;
+
+        let ctrlc = CtrlC::new()?;
+
+        let ctrlc_future = async {
+            ctrlc.await;
+            process::exit(1);
+        };
+
+        let socket_listener_future = async {
+            match listener.accept().await {
+                Ok((socket, ..)) => self.handle_connection(socket).await,
+                Err(e) => Err(Error::from(e)),
+            }
```

Comment:
I believe there's a loop missing from somewhere around here. The listener should handle multiple reads/writes to the socket, not just one.